Disease Refactor

Fox-McCloud opened the pull request:
Diseases are no longer processed by their own unique controller; instead, disease processing is handled entirely on the mob, itself (both disease stages and spreading).

There is still the disease SS, but this purely for holding info about various virus lists; it doesn't fire.

Went around and also made various virus lists on mobs istypeless for loops memes.

Another issue resolved was checking for necessary organs to contract a disease; this was done in disease `New`, and since `affected_mob` wouldn't even be around yet, well....nothing would ever happen. That's now properly handled in the `CanContractDisease` proc.

Added a new disease level var `bypasses_immunity` to bypass a species' `VIRUSIMMUNE` trait. Remember, `/datum/disease` is a system for ***both*** diseases and viruses. In the future, we may want someone to be able to acquire a disease, while still retaining immunity to viruses.

Lastly, re-ordered chemical processing and breathing in the `Life` loop to avoid some edge cases of circumventing breathing O2 damage with chems.


I tested infecting mobs, curing them, utilizing existing admin panels to cure, mixing viruses in blood, mutating viruses in blood, infecting mobs with said blood, renaming existing  advanced diseases, andddd tested to make sure spreading from mob to mob, passively still worked.

Other changes:
- TB bypasses species virus immunity (it's a condition associated with the lungs and is bacterial+fungal in nature; not a disease)
- Appendicitis bypasses species virus immunity; it's now entirely based on if you have an appendix or not
- Brainrot checks for the brain instead of the head; TB checks for lungs instead of the head
- Disease and Appendicitis events less likely to pick players who have left the round
- Fixes changeling panacea curing beneficial viruses

All seems to function as intended.

:cl: Fox McCloud
del: blood and gibs on turfs can no longer infect nearby people
tweak: Tuberculosis bypasses species virus immunity (it's not a virus!)
tweak: Disease and appendicitis events no longer infect clientless mobs
tweak: Disease event will no longer pick virus immune mobs
tweak: Appendicitis event will no longer pick mobs without an appendix
fix: Fixed changeling panacea curing non-harmful viruses
/:cl:

Fox-McCloud committed:
Vomit Now Contains Viruses

Fox-McCloud committed:
eh

Cyberboss commented:
I swear all blood had viruses

Durkel commented:
It was removed because it was causing a shit ton of lag, Is this going to cause the same problem?

Supermichael777 commented:
what caused the lag was every time a mob crossed a blood decal 
it checked if the blood had a disease
it checked whether the mob was a valid type
it checked the mobs bio armor
it checked for the snowflaked up disease kit (white shoes, gloves, sterile masks, internals, uniform, lab-coat)
it checked if it was disease immune
it rolled to see if they were infected.
not necessarily in an optimized order

Cyberboss commented on code/game/objects/effects/decals/cleanable/misc.dm:
```diff
@@ -82,7 +82,16 @@
-/obj/effect/decal/cleanable/vomit/attack_hand(var/mob/user)
+/obj/effect/decal/cleanable/vomit/Initialize(mapload, list/virus_list, virus_chance = 20)
+	. = ..()
+	if(virus_list && virus_list.len)
```
if(length(virus_list))

Cyberboss commented on code/game/objects/effects/decals/cleanable/misc.dm:
```diff
@@ -82,7 +82,16 @@
+/obj/effect/decal/cleanable/vomit/Initialize(mapload, list/virus_list, virus_chance = 20)
+	. = ..()
+	if(virus_list && virus_list.len)
+		for(var/datum/disease/D in virus_list)
```
typeless loop

Fox-McCloud replied:
I'll do it, but I'm still not convinced it actually saves worthwhile CPU cycles, in most cases.

Cyberboss commented on code/game/objects/effects/decals/cleanable/misc.dm:
```diff
@@ -96,6 +105,8 @@
 							H.nutrition += nutri_check.nutriment_factor * nutri_check.volume
 							reagents.remove_reagent(nutri_check.id,nutri_check.volume)
 			reagents.trans_to(H, reagents.total_volume)
+			for(var/datum/disease/D in viruses)
```
typeless loop

Fox-McCloud commented:
@Cyberboss 

Blood you extract from someone then dump on the ground does have viruses.

Hitting a mob to generate blood? No viruses. Making a blood trail of blood you dumped on the ground? None of it is going to have viruses.

Currently only gibbing will generate viruses, along with, as I mentioned blood dumping.

Fox-McCloud committed:
typeless memery

Fox-McCloud commented:
@Cyberboss 

Alternatively, I could just remove disease processing from literally everywhere but mobs, with the eventual goal of killing off the disease controller and just having mobs handle diseases/viruses from within their `Life` loops.

Cyberboss commented:
It pains me that that already isn't the case, so go ahead

On Jul 8, 2017 1:14 AM, "Fox McCloud" <notifications@github.com> wrote:

> @Cyberboss <https://github.com/cyberboss>
>
> Alternatively, I could just remove disease processing from literally
> everywhere but mobs, with the eventual goal of killing off the disease
> controller and just having mobs handle diseases/viruses from within their
> Life loops.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgstation/tgstation/pull/29130#issuecomment-313834600>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHywemYo5ajfJakJdRWfzI-Umz0t2qQZks5sLxA9gaJpZM4ORjMn>
> .
>

Fox-McCloud committed:
Disease Refactor

Iamgoofball commented on code/datums/diseases/_MobProcs.dm:
```diff
@@ -147,6 +147,11 @@
+	if(length(D.required_organs))
+		for(var/thing in D.required_organs)
+			if(!locate(thing) in bodyparts)
+				if(!locate(thing) in internal_organs)
```
i really really really really don't like a double locate

Cyberboss replied:
Should be if(!((locate(thing) in bodyparts) || (locate(thing) in internal_organs))

Fox-McCloud commented:
Welp, went ahead and did it @Cyberboss

Fox-McCloud commented on code/controllers/subsystem/disease.dm:
```diff
@@ -15,27 +12,5 @@ SUBSYSTEM_DEF(disease)
-			processing.Remove(thing)
-		if (MC_TICK_CHECK)
-			return
+	..("P:[active_diseases.len]")
```
I kept this just for debugging purposes; useful to see how many diseases, in total, are active in all mobs..also for admins.

Fox-McCloud commented on code/datums/diseases/_MobProcs.dm:
```diff
@@ -39,7 +39,7 @@
 		viruses += DD
 		DD.affected_mob = src
-		DD.holder = src
+		SSdisease.active_diseases += DD //Add it to the active diseases list, now that it's actually in a mob and being processed.
```
I didn't put this in `New` because it seemed redundant for a disease to be created, then it fails to contract, only for it to be remove from the list, again.

Fox-McCloud commented on code/datums/diseases/_disease.dm:
```diff
@@ -161,20 +137,6 @@
 
-
-/datum/disease/New()
-	if(required_organs && required_organs.len)
```
Baddddd

Fox-McCloud commented on code/modules/admin/secrets.dm:
```diff
@@ -149,8 +149,9 @@
 			if(choice == "Yes")
 				message_admins("[key_name_admin(usr)] has cured all diseases.")
-				for(var/datum/disease/D in SSdisease.processing)
-					D.cure(D)
```
Passing an argument of itself doesn't even make sense. Someone was drunk.

Fox-McCloud commented on code/modules/mob/mob_defines.dm:
```diff
@@ -112,7 +112,7 @@
 //List of active diseases
 
-	var/list/viruses = list() // replaces var/datum/disease/virus
+	var/list/viruses = list() // list of all diseases in a mob
```
I'm not going to do it in this PR, but this needs to be renamed `diseases` some day; people need to get it out of their head that the disease system is purely for viruses; it stifles development.

Fox-McCloud commented on code/datums/diseases/_disease.dm:
```diff
@@ -46,21 +46,23 @@
 	var/infectivity = 65
 	var/cure_chance = 8
 	var/carrier = 0 //If our host is only a carrier
+	var/bypasses_immunity = 0 //Does it skip species virus immunity check? Some things may diseases and not viruses
```
`FALSE`

Fox-McCloud committed:
tweaks

XDTM commented:
This is why i wanted my PR merged, i bet this is gonna conflict SO MUCH

Fox-McCloud commented:
@XDTM 

I barely touched advanced diseases.

XDTM commented:
I did touch some other stuff, i bet there's gonna be something, and every time i fix conflicts i spend an hour fixing stuff like deleting my whole PR.


That said, this is great!

kevinz000 commented:
...
..........
I'm probably a dumbass today but...
Diseases and viruses aren't the same thing??????????????????????????

XDTM commented:
They are, pretty much, but there's pseudodiseases like appendicitis or revenant's magic affliction which use the same system.

Fox-McCloud commented:
@kevinz000 

Broadly speaking, there's `/datum/disease` which covers things like spanish flu, GBS, TB, DNA spread, and the likes. This system could even be used for something like a medical condition of going into shock, prion disorders, addictions, or the likes; the system is well suited for it and it's easy to add new diseases.

Within that scope, a disease can be a disease, sickness, infection, or even virus. Beesease is a perfect example of a "disease". So is TB.



within the disease system, however, there's the "advanced disease" system that virology uses; the "advance disease" is a virus.

Cyberboss commented on code/datums/diseases/_MobProcs.dm:
```diff
@@ -147,6 +147,11 @@
-	if(dna && (VIRUSIMMUNE in dna.species.species_traits))
+	if(dna && (VIRUSIMMUNE in dna.species.species_traits) && !D.bypasses_immunity)
 		return 0
+	if(length(D.required_organs))
```
No need for the length check here

Cyberboss commented on code/datums/diseases/_disease.dm:
```diff
@@ -93,13 +95,16 @@
 
-/datum/disease/proc/spread(atom/source, force_spread = 0)
+
+/datum/disease/proc/spread(force_spread = 0)
```
FALSE

Fox-McCloud replied:
This isn't boolean. See a few lines below; it controls how far it will spread if `force_spread` is passed in an argument.

Cyberboss commented on code/datums/mind.dm:
```diff
@@ -631,8 +631,9 @@
-			for(var/datum/disease/D in current.viruses)
-				if(istype(D, /datum/disease/transformation/jungle_fever)) found = 1
+			for(var/datum/disease/transformation/jungle_fever/JF in current.viruses)
+				found = 1
```
TRUE

Cyberboss commented on code/datums/mind.dm:
```diff
@@ -631,8 +631,9 @@
 			text += "<a href='?src=\ref[src];monkey=healthy'>healthy</a>|<a href='?src=\ref[src];monkey=infected'>infected</a>|<b>HUMAN</b>|other"
 		else if (ismonkey(current))
 			var/found = 0
```
FALSE

Cyberboss commented on code/game/objects/effects/spawners/gibspawner.dm:
```diff
@@ -6,7 +6,7 @@
 
-/obj/effect/gibspawner/Initialize(mapload, list/viruses, datum/dna/MobDNA)
+/obj/effect/gibspawner/Initialize(mapload, datum/dna/MobDNA)
 	..()
```
. = ..()

Cyberboss commented on code/game/objects/items/weapons/melee/misc.dm:
```diff
@@ -148,7 +148,7 @@
 			H.internal_organs -= B
 			qdel(B)
-		new /obj/effect/gibspawner/generic(H.loc, H.viruses, H.dna)
+		new /obj/effect/gibspawner/generic(H.loc, H.dna)
```
H.loc => get_turf(H)

Cyberboss commented on code/modules/mob/living/carbon/alien/death.dm:
```diff
@@ -1,8 +1,8 @@
+		new /obj/effect/gibspawner/xeno(loc)
 	else
-		new /obj/effect/gibspawner/xenobodypartless(loc,viruses)
+		new /obj/effect/gibspawner/xenobodypartless(loc)
```
both get_turf(src)

Cyberboss commented on code/modules/mob/living/carbon/alien/larva/death.dm:
```diff
@@ -8,9 +8,9 @@
+		new /obj/effect/gibspawner/larva(loc)
 	else
-		new /obj/effect/gibspawner/larvabodypartless(loc,viruses)
+		new /obj/effect/gibspawner/larvabodypartless(loc)
```
both get_turf(src)

Cyberboss commented on code/modules/mob/living/carbon/human/death.dm:
```diff
@@ -6,9 +6,9 @@
+		new /obj/effect/gibspawner/human(loc, dna)
 	else
-		new /obj/effect/gibspawner/humanbodypartless(loc, viruses, dna)
+		new /obj/effect/gibspawner/humanbodypartless(loc, dna)
```
both get_turf(src)

Cyberboss commented on code/modules/mob/living/death.dm:
```diff
@@ -18,7 +18,7 @@
 
 /mob/living/proc/spawn_gibs()
-	new /obj/effect/gibspawner/generic(loc, viruses)
+	new /obj/effect/gibspawner/generic(loc)
```
get_turf(src)

Cyberboss commented on code/modules/mob/living/silicon/death.dm:
```diff
@@ -1,5 +1,5 @@
 /mob/living/silicon/spawn_gibs()
-	new /obj/effect/gibspawner/robot(loc,viruses)
+	new /obj/effect/gibspawner/robot(loc)
```
get_turf

Fox-McCloud committed:
styling

Fox-McCloud commented:
@Cyberboss 

Addressed all but one of your things; also went ahead and fixed up a few more `TRUE` and `FALSE` that you missed in addition to a few more gibspawners utilizing `loc` instead of `get_turf`.

Also ensuring `. = ..()` is utilized with all the other gibspawners.

The single thing I did not make `FALSE` is because it's not a boolean.

XDTM commented:
Some questions:

- Does restricting to Life ticks mean that (potentially) diseases won't tick on dead people?
- Since Life ticks are every 2 seconds and disease ticks are 1 second, does this mean i have to adjust all the numbers?

Cyberboss commented:
@XDTM Life still ticks on dead mobs

On Jul 9, 2017 11:24 AM, "XDTM" <notifications@github.com> wrote:

> Some questions:
>
>    - Does restricting to Life ticks mean that (potentially) diseases
>    won't tick on dead people?
>    - Since Life ticks are every 2 seconds and disease ticks are 1 second,
>    does this mean i have to adjust all the numbers?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgstation/tgstation/pull/29130#issuecomment-313926482>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHywekuAaKuhpqZ4352BiMLGNFb53GOlks5sMPCygaJpZM4ORjMn>
> .
>

Fox-McCloud commented:
@XDTM 

No on both counts.

`handle_diseases` is called during the `Life` loop; as @XDTM stated, `Life` runs for dead mobs.

The spreading portion of the disease will run if they're alive or dead; the actual impacts the disease has on the mob will only run when they're alive.

As for how often a controller sicks; I believe you're misinformed here. `Life` does tick every `2` seconds, but so do diseases; they do not tick every second.

XDTM commented:
Could've sworn it was every second, but testing it's actually every two and i just had a very skewed concept of time.

XDTM commented:
This also means that when my PR was testmerged it was technically balanced to be half as effective as they currently are on live. Not sure if i should adjust or not.

Fox-McCloud committed:
tweaks

Fox-McCloud commented:
Made a few more tweaks; missed a couple of typeless for loops.

Also made the disease and appendix events less likely to hit a mob who's exited the round.

Made brainrot and TB checks for their respective organs (brain and lungs) instead of the head.

TB now bypasses species virus immunity...'cause it's not a virus.

Fox-McCloud commented:
@Cyberboss 

Anything else?

XDTM commented:
Virus immunity is also used by androids, which i think shouldn't be affected. VIRUSIMMUNE in general should include bacteria really, it's meant as disease immune.

Fox-McCloud commented:
There's a better solution for disease like that: just don't give them the necessary organ that would allow them to be infected.

Why would an Android have lungs, for instance? If they have lungs, then they could still be infected by something like a bacterial or fungal agent. That said, if they don't have lungs, then everything will be A-oK for them and they'll be immune to TB.

Appendicitis functions the same way.

XDTM commented:
NOBREATH should be handling that, yeah. Still, there's no real reason to separate bacterial diseases and viral diseases with VIRUSIMMUNE, since the immunity is always due to an inorganic or alien biology.

Fox-McCloud commented:
I'm going to respectfully disagree with that for a number of reasons.

If something has lungs or an appendix or is generally organic in nature, it's also going to be able to be infected by bacterial agents (even more so than viruses); viruses are usually species or host specific, but fungal and bacterial organisms tend to infect a much wider range of hosts.

Appendicitis, for instance, isn't even caused by bacteria; it's caused by a blockage.

Either case, I'd like a maintainers opinion on this further, and a potential review.

Fox-McCloud committed:
Merge remote-tracking branch 'upstream/master' into virus-vomit

Fox-McCloud committed:
Merge remote-tracking branch 'upstream/master' into virus-vomit

Fox-McCloud committed:
unecessary comment removal

Fox-McCloud commented:
@Cyberboss 

Conflicts resolved, again.

Fox-McCloud committed:
whoops

Fox-McCloud commented on code/modules/clothing/ears/ears.dm:
```diff
@@ -1,50 +1,50 @@
-
-//Ears: currently only used for headsets and earmuffs
```
This showed up when merge conflicts were resolved; I have no idea why; nothing was changed here.

Fox-McCloud committed:
Merge remote-tracking branch 'upstream/master' into virus-vomit

Fox-McCloud commented:
@Cyberboss 

Conflicts fixed. Getting tired of repeatedly resolving conflicts.

AnturK commented on code/modules/events/disease_outbreak.dm:
```diff
@@ -27,11 +27,17 @@
-		var/foundAlready = 0	// don't infect someone that already has the virus
-		for(var/datum/disease/D in H.viruses)
-			foundAlready = 1
+		if(!H.client)
```
I believe lack of this check was intentional, maybe not interesting but intended.

Cyberboss replied:
Eh, I'm fine with it

AnturK commented on code/modules/events/shuttle_loan.dm:
```diff
@@ -132,7 +132,7 @@
 
 			if(ANTIDOTE_NEEDED)
-				var/virus_type = pick(/datum/disease/beesease, /datum/disease/brainrot, /datum/disease/fluspanish)
+				var/obj/item/weapon/reagent_containers/glass/bottle/virus_type = pick(/obj/item/weapon/reagent_containers/glass/bottle/beesease, /obj/item/weapon/reagent_containers/glass/bottle/brainrot, /obj/item/weapon/reagent_containers/glass/bottle/fluspanish)
```
Kills the shuttle purpose in my opinion.

Fox-McCloud replied:
Yeah, it's the one thing I didn't like; I thought this little event was neat.

AnturK replied:
Some kind of snowflake spill subtype ?

Fox-McCloud replied:
I think a dead infected assistant would probably be the best thing; having smoke wouldn't really be that impressive since people would just avoid it.

If it's an assistant that's infected though, then people have to engage in proper containment procedures as if it were blood.

AnturK reviewed (approved):
Looks good, just some bikeshedding about content changes.

Iamgoofball commented on code/modules/surgery/organs/liver.dm:
```diff
@@ -1,66 +0,0 @@
-#define LIVER_DEFAULT_HEALTH 100 //amount of damage required for liver failure
```
uhhhhh

Fox-McCloud committed:
Merge remote-tracking branch 'upstream/master' into virus-vomit

Fox-McCloud commented on code/modules/events/shuttle_loan.dm:
```diff
@@ -21,7 +21,7 @@
 
 /datum/round_event/shuttle_loan/start()
-	dispatch_type = pick(HIJACK_SYNDIE, RUSKY_PARTY, SPIDER_GIFT, DEPARTMENT_RESUPPLY, ANTIDOTE_NEEDED, PIZZA_DELIVERY)
+	dispatch_type = ANTIDOTE_NEEDED /*pick(HIJACK_SYNDIE, RUSKY_PARTY, SPIDER_GIFT, DEPARTMENT_RESUPPLY, ANTIDOTE_NEEDED, PIZZA_DELIVERY) */
```
oops. Remove this debugging.

Fox-McCloud commented on code/modules/mob/living/life.dm:
```diff
@@ -57,6 +61,12 @@
+/mob/living/proc/handle_diseases()
+	return
+
+/mob/living/proc/handle_chemicals_in_body()
```
What. No.

Fox-McCloud committed:
fix

Fox-McCloud commented:
@AnturK @Cyberboss 

Merge conflicts resolved.
